File: frontend/src/services/authService.ts
import apiClient from './apiClient'

export interface LoginPayload { email: string; password: string }
export interface RegisterPayload {
  full_name: string; email: string; phone: string
  company_name: string; password: string; confirm_password: string
}
export interface TokenOut {
  access_token: string; token_type: string
  dealer_id: number; full_name: string; email: string
}
export interface DealerOut {
  id: number; full_name: string; email: string
  phone: string; company_name: string; address?: string; is_active: boolean
}
export interface DealerUpdate {
  full_name?: string; phone?: string; company_name?: string; address?: string
}
export interface PasswordChange {
  current_password: string; new_password: string; confirm_password: string
}

export const authService = {
  async login(payload: LoginPayload): Promise<TokenOut> {
    const form = new URLSearchParams()
    form.append('username', payload.email)
    form.append('password', payload.password)
    const { data } = await apiClient.post<TokenOut>('/auth/login', form, {
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    })
    return data
  },

  async register(payload: RegisterPayload): Promise<DealerOut> {
    const { data } = await apiClient.post<DealerOut>('/auth/register', payload)
    return data
  },

  async getMe(): Promise<DealerOut> {
    const { data } = await apiClient.get<DealerOut>('/auth/me')
    return data
  },

  async updateProfile(payload: DealerUpdate): Promise<DealerOut> {
    const { data } = await apiClient.put<DealerOut>('/dealers/me', payload)
    return data
  },

  async changePassword(payload: PasswordChange): Promise<{ message: string }> {
    const { data } = await apiClient.put('/dealers/me/password', payload)
    return data
  },
}

File: backend/app/services/auth_service.py
"""Authentication service — register, login, profile."""

from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.core.security import create_access_token, verify_password
from app.repositories.dealer_repo import DealerRepository
from app.schemas.dealer import DealerRegister, DealerLogin, DealerOut, DealerUpdate, PasswordChange, TokenOut


class AuthService:
    def __init__(self, db: Session):
        self.repo = DealerRepository(db)

    def register(self, data: DealerRegister) -> DealerOut:
        if self.repo.get_by_email(data.email):
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="An account with this email already exists.",
            )
        dealer = self.repo.create(data)
        return DealerOut.model_validate(dealer)

    def login(self, data: DealerLogin) -> TokenOut:
        dealer = self.repo.get_by_email(data.email)
        if dealer is None or not verify_password(data.password, dealer.hashed_password):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid email or password.",
            )
        if not dealer.is_active:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Your account is disabled. Contact support.",
            )
        token = create_access_token(subject=dealer.id)
        return TokenOut(
            access_token=token,
            dealer_id=dealer.id,
            full_name=dealer.full_name,
            email=dealer.email,
        )

    def update_profile(self, dealer_id: int, data: DealerUpdate) -> DealerOut:
        dealer = self.repo.get_by_id(dealer_id)
        if dealer is None:
            raise HTTPException(status_code=404, detail="Dealer not found")
        updated = self.repo.update(dealer, data)
        return DealerOut.model_validate(updated)

    def change_password(self, dealer_id: int, data: PasswordChange) -> dict:
        dealer = self.repo.get_by_id(dealer_id)
        if dealer is None:
            raise HTTPException(status_code=404, detail="Dealer not found")
        if not verify_password(data.current_password, dealer.hashed_password):
            raise HTTPException(status_code=400, detail="Current password is incorrect.")
        self.repo.update_password(dealer, data.new_password)
        return {"message": "Password updated successfully."}
